### PR TITLE
[opensprinkler] Fix Program names and add new features for firmware 2.2.0

### LINKED ISSUE #15260: [OpenSprinkler]- Program Channel not displaying properly OS Firmware v2.2
## Expected Behavior

With OpenSprinkler Firmware v2.19 Program names were properly displayed by the binding (3.4.x,4.0.x).  With Firmware v2.20 this is no longer the case.

## Current Behavior

Since updating Opensprinkler firmware from v2.19 to v2.2 the program channel no longer displays correctly.  Instead of the Program Name, 415.0] is displayed for each/all Programs.  From reviewing the API documentation for each version of the firmware, there appears to have been a change in JSON structure for how Programs are stored, so it appears as though the JSOn is no longer being parsed correctly.
 These examples were taken directly from the API Docs.

v2.2
Example Return:
{“nprogs”:3, “nboards”:1, “mnp”:40, “mnst”:4, “pnsize”:32,“pd”:[[3,127,0,[480,2,240,0],[0,2700,0,2700,0,0,0,0],“Summer”,[0,33,**415]**], [2,9,0,[120,0,300,0],[0,3720,0,0,0,0,0,0],“Fall Prog”,[0,33,**415]**],[195,16,0,[1150,-1,-1,-1],[0,0,0,0,0,0,64800,0],“Pipe”,[1,67,**415]**]]}

v2.19
Example Return​:
{“nprogs”:3, “nboards”:1, “mnp”:40, “mnst”:4, “pnsize”:32,“pd”:[[3,127,0,[480,2,240,0],[0,2700,0,2700,0,0,0,0],“Summer”], [2,9,0,[120,0,300,0],[0,3720,0,0,0,0,0,0],“Fall Prog”],[67,16,0,[1150,-1,-1,-1],[0,0,0,0,0,0,64800,0],“Pipe”]]}
## Steps to Reproduce (for Bugs)

1.
2.

## Context

OpenSprinkler Programs can no longer be differentiated and selected from OpenHAB UI.  No way to tell one program from thenext as they all have the same, incorrect, name.  **415.0]**

## Your Environment

 Running OpenHAB 4.0Mx in a Docker container on a Linux Server.  Same issue exists running OpenHAB 3.4.3 in Docker.

### PR DESCRIPTION
**Pull request by Skinah**
This change fixes the bindings ability to scrape the program names of the sprinkler controllers internal automated programs that a user can select to run. More information is found that the issue this fixes here closes #15260

The newer firmware uses JSON with this structure and the "pd" now has an extra element over previous firmware which causes the update to break the old code. The JSON is an Array that contains mixed types including further arrays. Details are included in case a better method exists.

```
{
	"nprogs": 3,
	"nboards": 1,
	"mnp": 40,
	"mnst": 4,
	"pnsize": 32,
	"pd": [
		[3, 68, 0, [20540, 0, 120, 0],
			[0, 2700, 2700, 0, 0, 0, 0, 0], Water Lawn, [0, 33, 415]
		],
		[3, 1, 0, [16384, 0, 0, 0],
			[0, 0, 0, 2700, 0, 0, 0, 0], Water Citrus, [0, 33, 415]
		],
		[3, 18, 0, [16384, 0, 0, 0],
			[2700, 0, 0, 0, 0, 0, 0, 0], Water Fruit Trees, [0, 33, 415]
		]
	]
}
```

This pull request will automatically be built and available under the following links if anyone wants to test:
https://openhab.jfrog.io/ui/native/libs-pullrequest-local/org/openhab/addons/bundles/

https://openhab.jfrog.io/ui/native/libs-pullrequest-local/org/openhab/addons/bundles/org.openhab.binding.opensprinkler/4.1.0-SNAPSHOT/org.openhab.binding.opensprinkler-4.1.0-SNAPSHOT.jar

Signed-off-by: Matthew Skinner <matt@pcmus.com>

### CODE REVIEW

File: bundles/org.openhab.binding.opensprinkler/pom.xml [hunk 1]
```diff
@@ -13,5 +13,12 @@
...
   <name>openHAB Add-ons :: Bundles :: OpenSprinkler Binding</name>
-
+  <dependencies>
+    <dependency>
+      <groupId>javax.measure</groupId>
+      <artifactId>unit-api</artifactId>
+      <version>2.2</version>
+      <scope>compile</scope>
+    </dependency>
+  </dependencies>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.opensprinkler/pom.xml:23]: Why is it necessary to add this dependency ? It is provided by default ?

File: bundles/org.openhab.binding.opensprinkler/src/main/feature/feature.xml [hunk 2]
```diff
@@ -3,6 +3,7 @@
 	<repository>mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/${ohc.version}/xml/features</repository>
 	<feature name="openhab-binding-opensprinkler" description="OpenSprinkler Binding" version="${project.version}">
 		<feature>openhab-runtime-base</feature>
+		<feature dependency="true">javax.measure.unit-api</feature>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.opensprinkler/src/main/feature/feature.xml:6]: Why is it necessary to add this dependency ?

File: bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml [hunk 3]
```diff
@@ -123,6 +134,13 @@
 		<state readOnly="true"/>
 	</channel-type>
 
+	<channel-type id="queuedZones">
+		<item-type>Number:Dimensionless</item-type>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml:138]: Why using `Number:Dimensionless`?
It is a simple number, no ?

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerApi.java [hunk 4]
```diff
@@ -248,4 +248,34 @@ public interface OpenSprinklerApi {
...
+     * Returns the paused status of the OpenSprinkler.
+     *
+     * @return int 0 to 600 seconds
+     */
+    int getPausedState();
+
+    /**
+     * Sets the amount of time that the OpenSprinkler will stop/pause zones.
+     *
+     * @param 0 to 600 seconds
```
Comment (lolodomo) [inline bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerApi.java:276]:
```
@param seconds the pause duration in seconds (0 to 600)
```

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/handler/OpenSprinklerDeviceHandler.java [hunk 5]
```diff
@@ -116,6 +116,19 @@ protected void updateChannel(ChannelUID channel) {
                 break;
             case CHANNEL_RESET_STATIONS:
                 break;
+            case CHANNEL_QUEUED_ZONES:
+                updateState(channel, new QuantityType<Dimensionless>(localAPI.getQueuedZones(), Units.ONE));
```
Comment (lolodomo) [inline bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/handler/OpenSprinklerDeviceHandler.java:120]: No reason to use a QuantityType here, your state is a simple number.

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/OpenSprinklerState.java [hunk 6]
```diff
@@ -63,6 +63,11 @@ public static class JcResponse {
         public int rssi = 1;
         public int flcrt = -1;
         public int curr = -1;
+        public int pq = -1;
```
Comment (lolodomo) [inline bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/OpenSprinklerState.java:66]: Seems to be never used.

File: bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml [hunk 7]
```diff
@@ -89,6 +89,9 @@
 			<channel id="resetStations" typeId="resetStations"></channel>
 			<channel id="enablePrograms" typeId="enablePrograms"></channel>
 			<channel id="rainDelay" typeId="rainDelay"></channel>
+			<channel id="queuedZones" typeId="queuedZones"></channel>
+			<channel id="cloudConnected" typeId="cloudConnected"></channel>
+			<channel id="pausePrograms" typeId="pausePrograms"></channel>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml:94]: Please provide thing upgrade instructions (for an automatic adding of these new channels to existing things created with Main UI).

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.opensprinkler/README.md [hunk 8]
```diff
@@ -61,22 +61,25 @@ NOTE: Some channels will only show up if the hardware has the required sensor an
 
 | Channel Type ID | Item Type              |    | Description                                                                        |
 |-----------------|------------------------|----|------------------------------------------------------------------------------------|
-| rainsensor      | Switch                 | RO | This channel indicates whether rain is detected by the device or not.              |
-| sensor2         | Switch                 | RO | This channel is for the second sensor (if your hardware supports it).              |
+| cloudConnected  | Switch                 | RO | If the device is fully connected to the OpenSprinkler cloud this will show as 'ON'.|
```
Comment (lolodomo) [inline bundles/org.openhab.binding.opensprinkler/README.md:64]: I am asking myself if a channel is really the good option.
The connection status should probably be something defining the thing status ?

Comment (lolodomo) [reply bundles/org.openhab.binding.opensprinkler/README.md:64]: @jlaur : WDYT ?

Comment (jlaur) [reply bundles/org.openhab.binding.opensprinkler/README.md:64]: I think I need some more context. This binding a declared as `local`, so if I understand this correctly, this is just some information from the device whether it is also integrated with a cloud service in addition to being integrated through the local API? In that case, what would be the use of this channel? Perhaps it would be sufficient to provide this as a property?

File: bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml (hunk 3 above)
Comment (Skinah) [reply bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml:138]: I have no reason. What is Number:Dimensionless for then and when would be be used over just Number?

File: bundles/org.openhab.binding.opensprinkler/README.md (hunk 8 above)
Comment (Skinah) [reply bundles/org.openhab.binding.opensprinkler/README.md:64]: I believe a property is wrong as it has a state that changes. Read only channel seemed to fit. Your thoughts are correct and I also do not see the use case, but was easy to add it when doing other channels instead of someone requesting it later. It could possibly be used to fault find why you can not use their cloud to connect and control the sprinkler. The binding is full local and it is perhaps strange that the device can do both without trying to force you to do one over the other.

Comment (Skinah) [reply bundles/org.openhab.binding.opensprinkler/README.md:64]: I would prefer to delete the channel then to spend any more time on changing or adding extra lines to add it to things that are missing the channel. I don't have a lot of spare time these days.

File: bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml (hunk 3 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml:138]: When your value has a unit but no defined dimension like for example percentage or decibel.

File: bundles/org.openhab.binding.opensprinkler/pom.xml (hunk 1 above)
Comment (Skinah) [reply bundles/org.openhab.binding.opensprinkler/pom.xml]: I was getting errors when trying to build, but I just rebased and removed this and all works without this now, so have it removed. Marking as solved.

File: bundles/org.openhab.binding.opensprinkler/src/main/feature/feature.xml (hunk 2 above)
Comment (Skinah) [reply bundles/org.openhab.binding.opensprinkler/src/main/feature/feature.xml]: Removed and it works fine without.

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerApi.java (hunk 4 above)
Comment (Skinah) [reply bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerApi.java]: applied suggestion.

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV220.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    @Override
+    public void getProgramData() throws CommunicationApiException, UnauthorizedApiException {
+        String returnContent;
+        try {
+            returnContent = http.sendHttpGet(getBaseUrl() + CMD_PROGRAM_DATA, getRequestRequiredOptions());
+        } catch (CommunicationApiException exp) {
+            throw new CommunicationApiException(
+                    "There was a problem in the HTTP communication with the OpenSprinkler API: " + exp.getMessage());
+        }
+        JpResponse resp = gson.fromJson(returnContent, JpResponse.class);
```
Comment (lolodomo) [inline bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV220.java:51]: Catch the possible exceptions (JsonParseException)

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/handler/OpenSprinklerDeviceHandler.java [hunk 10]
```diff
@@ -116,6 +116,19 @@ protected void updateChannel(ChannelUID channel) {
...
                 break;
+            case CHANNEL_QUEUED_ZONES:
+                updateState(channel, new DecimalType(localAPI.getQueuedZones()));
+                break;
+            case CHANNEL_CLOUD_CONNECTED:
+                if (localAPI.getCloudConnected() == 3) {
+                    updateState(channel, OnOffType.ON);
+                } else {
+                    updateState(channel, OnOffType.OFF);
+                }
```
Comment (lolodomo) [inline bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/handler/OpenSprinklerDeviceHandler.java:127]: Just a suggestion: it can be reduced to:
```
updateState(channel, OnOffType.valueOf(localAPI.getCloudConnected() == 3));
```

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/handler/OpenSprinklerDeviceHandler.java [hunk 11]
```diff
@@ -233,6 +258,24 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
+                            api.setPausePrograms(0);
+                        } else if (command instanceof DecimalType) {
+                            api.setPausePrograms(((BigDecimal) command).intValue());
+                        } else if (command instanceof QuantityType<?>) {
+                            QuantityType<?> quantity = (QuantityType<?>) command;
+                            quantity = quantity.toUnit(Units.SECOND);
+                            if (quantity != null) {
+                                api.setPausePrograms(quantity.toBigDecimal().intValue());
+                            }
+                            api.setPausePrograms(((BigDecimal) command).intValue());
```
Comment (lolodomo) [inline bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/handler/OpenSprinklerDeviceHandler.java:272]: Looks like you forget to remove this line.

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.opensprinkler/README.md [hunk 12]
```diff
@@ -61,22 +61,25 @@ NOTE: Some channels will only show up if the hardware has the required sensor an
...
-| nextDuration    | Number:Time            | RW | The time the station will open for when any stations are selected from the         |
-|                 |                        |    | `stations` channel. Defaults to 30 minutes if not set.                           |
-| resetStations   | Switch                 | RW | The ON command will stop all stations immediately, including those waiting to run. |
-| enablePrograms  | Switch                 | RW | Allow programs to auto run. When OFF, manually started stations will still work.   |
-| rainDelay       | Number:Time            | RW | Sets/Shows the amount of time (hours) that rain has caused programs to be delayed. |
+| waterlevel      | Number:Dimensionless   | RO | This channel shows the current water level in percent (0-250%). The water level is |
+|                 |                        |    | calculated based on the weather and influences the duration of the water programs. |
+| queuedZones     | Number:Dimensionless   | RO | A count of how many zones are running and also waiting to run in the queue.        |
```
Comment (lolodomo) [inline bundles/org.openhab.binding.opensprinkler/README.md:82]: Just Number

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/handler/OpenSprinklerDeviceHandler.java (hunk 10 above)
Comment (Skinah) [reply bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/handler/OpenSprinklerDeviceHandler.java:127]: I tried that but "OnOffType.valueOf(" does not support a boolean. It would be a good idea to extend/overload OnOffType.valueOf to support doing that.

Comment (Skinah) [reply bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/handler/OpenSprinklerDeviceHandler.java:127]: public static [OnOffType](https://www.openhab.org/javadoc/latest/org/openhab/core/library/types/onofftype) from(boolean state)

Used your suggestion but with .from(boolean)

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV220.java (hunk 9 above)
Comment (Skinah) [reply bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/api/OpenSprinklerHttpApiV220.java:51]: Done, thanks.

**Comment by Skinah** [comment]
@lolodomo I believe all changes are now addressed. Can you take a look over the update instructions to ensure I did them correctly please?

File: bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/handler/OpenSprinklerDeviceHandler.java (hunk 10 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/handler/OpenSprinklerDeviceHandler.java]: Sorry for mentioning the wrong method.

File: bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml [hunk 13]
```diff
@@ -60,6 +60,10 @@
 			<channel id="ignoreRain" typeId="ignoreRain"></channel>
 		</channels>
 
+		<properties>
+			<property name="thingTypeVersion">1</property>
+		</properties>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml:65]: Why did you add `thingTypeVersion` property to the station thing type while this thing type is not updated ?

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml (hunk 13 above)
Comment (Skinah) [reply bundles/org.openhab.binding.opensprinkler/src/main/resources/OH-INF/thing/thing-types.xml]: No reason, so I have removed your suggestion.

**Review by lolodomo — APPROVED** [review]
LGTM, thank you
